### DISCUSSION TITLE
Improve type definitions for the extensions sdk

### DISCUSSION THREAD

**Discussion by mattiaz9** [directus/directus#20670] [Feature Requests]
### Summary

Right now it's filled with `any` types which is frustrating and defies the purpose of having type definitions in the first place.

### Basic Example

- `useStores()`: it's defined as a `Record<string, any>` which means I have no idea how to use it, unless I go on github and look at the code
- api extensions `context.services` is a any type
- endpoint `request` parameter is missing some key additional props such as `accountability`

there are probably many more but those are the ones I use more often.

### Motivation

First of all it slows a tons the development and it makes it frustrating to work with. 
Also last time I checked those values are not even documented.

### Detailed Design

- Add here: https://github.com/directus/directus/blob/ae31ef8ac7d2e0451cc429d3158923992133a5c5/packages/composables/src/use-system.ts#L8 the definitions of this folder https://github.com/directus/directus/tree/ae31ef8ac7d2e0451cc429d3158923992133a5c5/app/src/stores
- Add this definitions: https://github.com/directus/directus/blob/7f017f1fc8141bcde34509d096b3a8bac499eed9/api/src/services/index.ts for the `context.services` type
- Create a custom `DirectusRequest` type with the extra props added by directus and use it in the `DefineEndpoint` router

### Must Have:
- `useStores` types
-  `context.serivces` types
- `request` additional props

### Drawbacks

I believe this should be fairly simple since most the types needed can be inferred 

### Alternatives

-

### Adoption Strategy

-

**Comment by northern-mindset** [comment]
Typing of the request object was mentioned a few times earlier:

- https://github.com/directus/directus/issues/10658#issuecomment-1002981899
- https://github.com/directus/directus/discussions/15098#discussioncomment-3642658

Mentioning it here for better discoverability.

A crazy idea: to move those `schema`, `accountability` and other extra fields from `req` to `context (ApiExtensionContext)`. 👀

**Reply by Vincent-HD** [reply]
For now I just augmented the interface like this to add the type, you can find more inside the discussion. Not ideal tho

```typescript
// index.d.ts
import { Accountability } from "@directus/types"

declare global {
    namespace Express {
        interface Request {
            accountability: Accountability
        }
    }
}
```
Here for the rest
https://github.com/directus/directus/issues/10658#issuecomment-1121534138

**Reply by rahadur** [reply]
> For now I just augmented the interface like this to add the type, you can find more inside the discussion. Not ideal tho
> 
> ```ts
> // index.d.ts
> import { Accountability } from "@directus/types"
> 
> declare global {
>     namespace Express {
>         interface Request {
>             accountability: Accountability
>         }
>     }
> }
> ```
> 
> Here for the rest [#10658 (comment)](https://github.com/directus/directus/issues/10658#issuecomment-1121534138)

Thanks it's worked for me..

**Comment by tonikny** [comment]
I can't find any doc to type context.services, is it a way to type ItemsService? Maybe with a generic or from table defined fields?
I'm using directus in a docker container.

**Reply by TBG-FR** [reply]
Using https://github.com/linefusion/indirectus I managed to get typed ItemsService, you might give it a try

**Comment by martin-braun** [comment]
<img width="841" alt="image" src="https://github.com/user-attachments/assets/7135ae6e-1615-40a9-ab92-dddba9492146">
This is such a frustrating experience. The first time I need to use `req.accountability`.

Yes, @rijkvanzanten, we need https://github.com/directus/directus/blob/51e90f6c4cbd1c3e9cd9e2b000eef694812280d5/api/src/types/express.d.ts#L1-L23 in the extension sdk as you mentioned here https://github.com/directus/directus/issues/10658#issuecomment-1121534138.

Can we actually do something about it?

I `npm i @directus/shared -D` and placed this in my project:

```ts
/**
 * Custom properties on the req object in express
 */

import { Accountability } from '@directus/shared/types';
import { Query } from '@directus/shared/types';

export type SchemaOverview = {
	[table: string]: {
		primary: string;
		columns: {
			[column: string]: {
				table_name: string;
				column_name: string;
				default_value: string | null;
				is_nullable: boolean;
				is_generated: boolean;
				data_type: string;
				numeric_precision?: number | null;
				numeric_scale?: number | null;
				max_length: number | null;
			};
		};
	};
};

declare global {
	namespace Express {
		export interface Request {
			token: string | null;
			collection: string;
			sanitizedQuery: Query;
			schema: SchemaOverview;

			accountability?: Accountability;
			singleton?: boolean;
		}
	}
}
```

It solves the issue. I see no reason to not have this the extension sdk. Typing support for custom extensions is still a very terrible experience, because of reasons like this.

**Reply by 7oque** [reply]
Don't know if helpful or not, but I am using `@directus/types` in my extension and then just copying the definition;

```typescript
/**
 * Custom properties on the req object in express
 */

import { Accountability, Query, SchemaOverview } from '@directus/types';

declare global {
	namespace Express {
		export interface Request {
			token: string | null;
			collection: string;
			sanitizedQuery: Query;
			schema: SchemaOverview;
			accountability?: Accountability;
			singleton?: boolean;
		}
	}
}
```
